### PR TITLE
POC flex gemm

### PR DESCRIPTION
**Pull request by drisspg**
## Add GEMM epilogue fusion prototype

This PR prototypes a GEMM epilogue fusion higher-order op plus Inductor lowering/codegen paths for fusing pointwise epilogues into GEMM-family ops.

### What is in this PR

- Adds `gemm_epilogue_fusion` and convenience wrappers for `mm`, `addmm`, `bmm`, `baddbmm`, grouped-mm, scaled-mm, and GEMM-shaped `matmul` epilogue fusion.
- Adds Dynamo/ProxyTensor/Inductor plumbing so the epilogue callable can be traced and lowered with the GEMM op.
- Adds TRITON coverage for common GEMM epilogue paths, including `_scaled_mm + epilogue`.
- Adds QUACK backend routing/codegen for `mm`, `addmm`, `bmm`, `baddbmm`, and MXFP8 blockscaled `_scaled_mm` epilogues.
- Adds CUTLASS/CUTEDSL backend acceptance/routing hooks; CUTLASS runtime validation still needs a proper CUDA/CUTLASS environment.
- Adds targeted tests in `test/inductor/test_gemm_epilogue_fusion.py`.

### API note

The generic form now supports passing the user-facing shorthand callable directly:

```py
gemm_epilogue_fusion(torch.mm, (x, y), epilogue_fn, kernel_options={"backend": backend})
```

The experiment below intentionally uses `torch.mm` instead of `torch.ops.aten.mm.default` to exercise that support.

### Current limitations / follow-ups

- QUACK generic epilogue bridge currently supports pointwise epilogues, not reductions such as `acc / acc.sum(dim=1, keepdim=True)`.
- QUACK `_scaled_mm` currently handles the MXFP8 blockscaled path; tensorwise float32 scales, bias, `scale_result`, and `use_fast_accum=True` are not wired yet.
- `matmul_epilogue` is limited to 2D `mm` and 3D `bmm` shaped inputs.
- `torch.nn.functional.linear` convenience wrapping is intentionally left out for now.

QuACK companion work: https://github.com/drisspg/quack/pull/new/make-epilogue-generic

### Repro / codegen experiment

```py
#!/usr/bin/env python3
import re
from pathlib import Path
from typing import Annotated

import torch
import torch._inductor.config as inductor_config
import typer
from torch._higher_order_ops import gemm_epilogue_fusion
from torch._inductor.utils import run_and_get_code
from transformer_nuggets.utils.benchmark import profiler


DTYPES = {"float16": torch.float16, "bfloat16": torch.bfloat16, "float32": torch.float32}
EPILOGUES = {
    "identity": lambda acc: acc,
    "relu": lambda acc: acc.relu(),
    "affine-relu": lambda acc: (acc * 0.5 + 1.0).relu(),
    "gelu": lambda acc: torch.nn.functional.gelu(acc),
    "silu": lambda acc: torch.nn.functional.silu(acc),
    "sigmoid": lambda acc: torch.sigmoid(acc),
    "tanh": lambda acc: torch.tanh(acc),
    "exp": lambda acc: torch.exp(acc),
    "abs": lambda acc: torch.abs(acc),
    "square": lambda acc: acc * acc,
    "clamp": lambda acc: acc.clamp(min=0.0, max=6.0),
    "leaky-relu": lambda acc: torch.nn.functional.leaky_relu(acc, 0.01),
    "where-leaky-relu": lambda acc: torch.where(acc > 0, acc, acc * 0.01),
    "sqrt-abs": lambda acc: torch.sqrt(torch.abs(acc) + 1.0),
    "rsqrt-abs": lambda acc: torch.rsqrt(torch.abs(acc) + 1.0),
    "normalize-sum-n": lambda acc: acc / acc.sum(dim=1, keepdim=True),
}


def write_text(path: Path, text: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text)


def main(
    out_dir: Annotated[
        Path,
        typer.Option(help="Directory for exported graph and generated code artifacts."),
    ] = Path(__file__).resolve().parent / "gemm_epilogue_codegen",
    m: Annotated[int, typer.Option(help="Rows of A/output.")] = 2048,
    n: Annotated[int, typer.Option(help="Columns of B/output.")] = 2048,
    k: Annotated[int, typer.Option(help="Reduction dimension.")] = 2048,
    dtype: Annotated[
        str,
        typer.Option(help="Input dtype.", case_sensitive=False),
    ] = "float16",
    epilogue: Annotated[
        str,
        typer.Option(help="Epilogue to try; see EPILOGUES in this script."),
    ] = "relu",
    backend: Annotated[
        str,
        typer.Option(help="Requested GEMM epilogue backend: TRITON, CUTEDSL, or QUACK."),
    ] = "QUACK",
    profile: Annotated[bool, typer.Option(help="Whether to run the compiled code with a profiler.")] = False, 
) -> None:
    if not torch.cuda.is_available():
        raise RuntimeError("This experiment expects CUDA to be available")
    if dtype not in DTYPES:
        raise typer.BadParameter(f"dtype must be one of {', '.join(DTYPES)}")
    if epilogue not in EPILOGUES:
        raise typer.BadParameter(f"epilogue must be one of {', '.join(EPILOGUES)}")

    torch.manual_seed(0)
    a = torch.randn(m, k, device="cuda", dtype=DTYPES[dtype])
    b = torch.randn(k, n, device="cuda", dtype=DTYPES[dtype])
    epilogue_fn = EPILOGUES[epilogue]

    def fn(x, y):
        return gemm_epilogue_fusion(
            torch.mm,
            (x, y),
            epilogue_fn,
            kernel_options={"backend": backend},
        )

    out_dir.mkdir(parents=True, exist_ok=True)
    eager = epilogue_fn(torch.mm(a, b))

    compiled = torch.compile(fn, backend="inductor", fullgraph=True)
    with inductor_config.patch({"max_autotune_gemm": False}):
        actual, codes = run_and_get_code(compiled, a, b)
    torch.testing.assert_close(actual, eager)

    kernels = []
    for i, code in enumerate(codes):
        write_text(out_dir / f"inductor_output_{i}.py", code)
        kernels.extend(
            re.findall(
                r"async_compile\.triton\([^\n]+, '''\n(.*?)\n'''",
                code,
                re.DOTALL,
            )
        )

    for i, kernel in enumerate(kernels):
        write_text(out_dir / f"triton_kernel_{i}.py", kernel)

    summary = [
        "FlexJam GEMM epilogue fusion experiment",
        f"torch: {torch.__file__}",
        f"git_version: {torch.version.git_version}",
        f"shape: ({m}, {k}) x ({k}, {n})",
        f"dtype: {dtype}",
        f"epilogue: {epilogue}",
        f"backend: {backend}",
        f"inductor_outputs: {len(codes)}",
    ]
    summary.extend(f"  - {out_dir / f'inductor_output_{i}.py'}" for i in range(len(codes)))
    summary.append(f"triton_kernels: {len(kernels)}")
    summary.extend(f"  - {out_dir / f'triton_kernel_{i}.py'}" for i in range(len(kernels)))
    summary.append(f"max_abs_diff: {(actual - eager).abs().max().item()}")

    write_text(out_dir / "summary.txt", "\n".join(summary) + "\n")
    print("\n".join(summary))

    if profile:
        print("\nProfiling the compiled code...")
        with profiler(f"gemm_epilogue_{epilogue}_backend_{backend}") as prof:
            compiled(a, b)


if __name__ == "__main__":
    typer.run(main)
```

Some profilign to make sure it is working for relu
image

### CODE REVIEW

**Comment by drisspg** [comment]
Closing: this POC has been superseded by the FlexGEMM vendor upstreaming stack (#188468, #188469, #188470, #188112, #188739).
